security/pkg/nodeagent/cache/secretcache.go

```diff
@@ -570,33 +570,34 @@ func (sc *SecretCache) rotate(updateRootFlag bool) {
 			return true
 		}
 
-		if sc.configOptions.CredFetcher != nil {
-			// Refresh token through credential fetcher.
-			cacheLog.Debugf("%s getting a new token through credential fetcher", logPrefix)
-			t, err := sc.configOptions.CredFetcher.GetPlatformCredential()
-			if err != nil {
-				cacheLog.Warnf("%s credential fetcher failed to get a new token, continue using the original token: %v", logPrefix, err)
-			} else {
-				secret.Token = t
-			}
-		}
-
```

Comment:
Please do not move this code block. This is going to break STS flow for VM.

Reply:
How will it break it ? Don't we have test cases ? 

It doesn't make sense to fetch a token every minute - when the token is not going to be used. If anyhting in VM requires the token to be fetched without being used - it should be fixed.

Reply:
Unfortunately we don't have good integration test framework for VM at the moment. I recently fixed the issue. Please do not revert the fix. It's fetched for a fix rotation period, definitely not every minute. I plan to work on further improving the performance based on token expiration period (which varies for each provider).

Reply:
Also this code should not be called if 'useCertToRotate'.

Reply:
It's unrelated to cert provisioning flow. Remember STS is using the same token. We need to refresh token regardless 'useCertToRotate'.

Reply:
> Unfortunately we don't have good integration test framework for VM at the moment.

Why? We have plenty of VM tests.

Reply:
The current test framework uses a k8s pod to simulate VM. It will never run into issues like no token or token expiration. We can improve on that. But right now there is no test coverage for the issue I discovered.

Reply:
It is no different than a real VM in terms of token. We do not mount any tokens for k8s - it is all done 100% manually just like a VM.

Reply:
OK cool. We can probably replace the bootstrap token flow with a mocked credential fetcher in some test cases. This can be a TODO for future VM test improvement.

Reply:
I still don't understand - if STS flow needs a token, it should get it - depending on a side-effect of this is not right.

Happy to make the proper fix - and add the missing comments - if you can explain what is supposed to happen. 
This is in master - we should do the right thing and cleanup the mess.

Reply:
The method is called "rotate" - it is not supposed to have side-effects for other parts of the code.

Reply:
I checked and Token field ( the only side effect of the method ) doesn't appear to be used in STS, which gets the token from Envoy, freshly read from disk.

Reply:
The token is used by Envoy. It's not in Istio agent code. Envoy's gRPC library is looking for the token that is set by istio agent. I already tested it end-to-end. The current code works perfectly. If we move it inside "shouldRotate", things will break as soon as token expires.

Reply:
Oh, I finally got it - so the rotate function has a side-effect of calling get token ( even if it doesn't need a token, because it's not rotating the tokens). The function to get a token has a side-effect on saving a file - which is in turn used by Envoy. 

I'll move the code and add a comment and TODO - but this is completely wrong to have in security code with no comments.

We will need to evaluate what to do with this - I don't know how this got past review in the first place, but it's really not right.

Reply:
Ok, moved and added the comment.